### PR TITLE
Purge the search tables directly

### LINKED ISSUE #2162: Non-existent service "contao.search.indexer"
**Affected version(s)**

Contao 4.10.0

**Description**

I've deactivated the search indexer in my Contao 4.9.5 Installation via `config.yml`:

```html
contao:
    search:
        default_indexer:
            enable: false
```

After update to 4.10.0 with Contao Manager the following error occurs:

```html
In CheckExceptionOnInvalidReferenceBehaviorPass.php line 86:

The service "Contao\CoreBundle\EventListener\DataContainer\PageUrlListener" 
has a dependency on a non-existent service "contao.search.indexer". 
```

After making a repair via Contao Manager I'm back on 4.9.5.

I activate the search indexer in my `config.yml` and run the update to 4.10 again without any problems.

### PR DESCRIPTION
**Pull request by aschempp**
| Q                | A
| -----------------| ---
| Fixed issues     | Fixes #2162

It does not make sense to delete from the search indexer(s), because we query the `tl_search` table directly. This can only work with the default indexer anyway.

### CODE REVIEW

**Review by leofeyer — APPROVED** [review]
Makes sense to me. 👍🏻 Let's wait for @Toflar's approval though.

**Review by Toflar — DISMISSED** [review]
Fine for me, however, it was @ausi who built that when implementing the new search :)

**Comment by ausi** [comment]
Shouldn’t we instead not query the `tl_search` table? 
How should this work with a non-default indexer?

> it was @ausi who built that when implementing the new search :)

I think I mainly switched to the indexer to remove the duplicated SQL code in [`6fe7dc9` (#1679)](https://github.com/contao/contao/pull/1679/commits/6fe7dc90683ff6c251eba12e99d80cc5890aa68a)

**Comment by leofeyer** [comment]
> Shouldn’t we instead not query the `tl_search` table?

How would we find the URL that belongs to the page ID then?

**Comment by aschempp** [comment]
It doesn't work with other indexes. The (might) need to register their own callbacks.

**Comment by ausi** [comment]
> How would we find the URL that belongs to the page ID then?

We could let the indexers themself decide what to delete?  
The page ID can be passed via JSON LD?

```php
$this->searchIndexer->delete(new Document(
	new Uri(''), 
	404, 
	[], 
	'<script type="application/ld+json">' . json_encode([
		'@context' => 'https://contao.org/',
		'@type' => 'Page',
		'pageId' => (int) $pageId,
	]) . '</script>'
));
```

**Comment by leofeyer** [comment]
That is a clever idea for sure. 🦊

**Comment by Toflar** [comment]
I propsed the same idea to @aschempp yesterday but there's just so much more that could possibly be added there. I can add my own JSON-LD data in the FE and use that to index. It would be missing here again.

**Comment by ausi** [comment]
The page ID is the only information we have in this case. If you added your own JSON-LD data, you can create a callback and call `searchIndexer->delete()` with your data the same way.

But this way we have the benefit, that the indexer can decide if and how it wants to handle the deletion.

**Comment by Toflar** [comment]
I tend to agree, yes.

**Comment by aschempp** [comment]
It might work. I'm just not sure the (massive) overhead is really worth the effort of _theoretically_ have someone implement their own search index.

@Toflar and me also discussed that search results should use the same tags the reverse proxy does. This would allow to correctly purge the search from everywhere (e.g. when changing a news item).

**Comment by leofeyer** [comment]
Ok, so how do we proceed? Are we going with @ausi's [suggestion above](https://github.com/contao/contao/pull/2265#issuecomment-690140320) or do I merge this PR?

> search results should use the same tags the reverse proxy does. This would allow to correctly purge the search from everywhere

I assume that this would be a second step in a separate feature PR.

**Comment by aschempp** [comment]
🤷 I would opt for my (simple) solution and hope for tags support in a future release …

**Comment by ausi** [comment]
> Ok, so how do we proceed?

I think we should merge this PR for now and create a new issue for the suggestion from https://github.com/contao/contao/pull/2265#issuecomment-690140320 
If someone finds the time (and/or needs the feature) they can implement it.

File: core-bundle/src/EventListener/DataContainer/PageUrlListener.php [hunk 1]
```diff
@@ -223,7 +215,9 @@ public function purgeSearchIndex(int $pageId): void
         ;
 
         foreach ($urls as $url) {
-            $this->searchIndexer->delete(new Document(new Uri($url), 200));
+            /** @var Search $search */
+            $search = $this->framework->getAdapter(Search::class);
```
Comment (leofeyer) [inline core-bundle/src/EventListener/DataContainer/PageUrlListener.php]: These two lines should be outside the `foreach` loop.

**Review by leofeyer — CHANGES_REQUESTED** [review]

File: core-bundle/src/EventListener/DataContainer/PageUrlListener.php (hunk 1 above)
Comment (aschempp) [reply core-bundle/src/EventListener/DataContainer/PageUrlListener.php]: not really related to this PR, but changed in 97e747ce07316253ed444fcbec25ca12884753da 😆

**Review by ausi — APPROVED** [review]

**Review by leofeyer — APPROVED** [review]

**Pull request merged**